simulators/main/event.py

```diff
@@ -0,0 +1,304 @@
+        self.last_production_time_update = datetime.timestamp(datetime.utcnow())
+        if self.machine_up and not self.shutdown:
+            self.production_time_s = self.production_time_s + production_time
+            log.debug(f'{self.device_id} production time: {self.production_time_s}s')
+
+    def pick_one_piece(self, pieces_per_seconds: float):
+        piece_production_time = 1.0 / pieces_per_seconds
+        if self.production_time_s > piece_production_time:
+            self.production_time_s = self.production_time_s - piece_production_time
+            log.debug(f'{self.device_id} one piece produced, remained time: {self.production_time_s}s')
+            return True
+        log.debug(f'{self.device_id} piece not yet produced, production time: {self.production_time_s}s')
+        return False
+
+    def pick_pieces(self, pieces_per_seconds: float):
+        pieces_produced = int(pieces_per_seconds * self.production_time_s)
+        self.production_time_s = self.production_time_s - pieces_produced / pieces_per_seconds
+        log.debug(f'{self.device_id} pieces produced: {pieces_produced}, remained time: {self.production_time_s}s')
+        return pieces_produced
+
+    def type_fragment(self, event_definition, text=None):
+        return {
+            'type': event_definition["type"],
+            'text': text or event_definition["type"]
+        }
+
+    def log_ignore(self, event_definition):
+        log.info(f'Device: {self.device_id} [{self.model["label"]}] is down -> ignore event {event_definition["type"]}')
+
+    def log_not_in_shift(self):
+        log.info(f'Device: {self.device_id} [{self.model["label"]}] is not in PRODUCTION shift -> ignore event')
+
+    def on_availability_event(self, event_definition, task):
+
+        self.produce_pieces()
+
+        event = self.type_fragment(event_definition)
+
+        statusses = event_definition.get("status") or ["up"]
+        probabilities = event_definition.get("probabilities") or [0.5]
+        durations = event_definition.get("durations") or [0]
+        status, duration = get_random_status(statusses, durations, probabilities)
+        self.machine_up = (status == "up")
+        event.update({'status': status})
+
+        event.update(self.get_production_info())
+        self.send_event(event)
+
+        # Send Piece_Produced as soon as possible(to get OEE calculation quicker for Slow Producers)
+        # Might make sense to configure the behavior per simulator in json.
+        if self.is_whole_piece_available():
+            self.force_production_event()
+
+        return duration
+
+    def force_production_event(self):
+        for event_definition in self.model["events"]:
+            event_type = event_definition.get("type")
+            if event_type == "Piece_Produced":
+                task = self.create_one_time_task(event_definition)
+                self.tasks.append(task)
```

Comment:
Is tasks present here?

Reply:
tasks are always created at the beginning so I can say they present here